Nest native serialization with class-validator/class-transformer

naz opened the pull request:
A spike for request data validation to be used in controllers

naz committed:
Basic pipe controller query input transformation

no issue

- Shows how the implementation of `formats`, `limit`, `page`, and `filter`  query parameters handling can be implemented using `pipes` transform

naz committed:
Added request body validation POST endpoint

- Adds nest's native validation for POST /snippets request body using input DTO and class-transformer / class-validator libraries

allouis commented on ghost/ghost/src/http/pipes/filter.pipe.ts:
```diff
@@ -0,0 +1,10 @@
+
+@Injectable()
+export class ParseFilterQueryPipe implements PipeTransform {
+    // @NOTE: the return type here should ideally be a typed NQL object (ask Egg why!)
```
😂

allouis commented on ghost/ghost/src/http/controllers/snippet.dto.input.ts:
```diff
@@ -0,0 +1,26 @@
+
+    @IsString()
+    @IsOptional()
+        lexical?: string;
```
It's annoying that `lexical?: string` literally means it's a string and it's optional, but we have to add decorators for both of these things 😕

naz replied:
Yeah... without "IsOptional" it failed the "IsString" validation. Quite verbose this validator is

naz committed:
Renamed output dto file to match the convention

naz committed:
Added browse endpoint DTO-based serialization

- The NestJS recommended way using "serializers" is crossing the architectural boundary into the business layer's Entities.
- This change demonstrates how we can utilize DTO objects per endpoint to achieve "allowlist" effect of the API output and add extra mapping logic as needed

naz committed:
Added validation in snippets
controller

- Allows to validate PUT requests to /admin/snippets endpoint using SnippetsBodyDTO with class-validator
